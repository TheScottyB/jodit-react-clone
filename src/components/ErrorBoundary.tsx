import React, { Component, ErrorInfo, ReactNode } from 'react';

interface ErrorBoundaryProps {
  /**
   * Child components that the error boundary wraps
   */
  children: ReactNode;
  
  /**
   * Optional custom fallback UI to render when an error occurs
   * If not provided, a default error message will be shown
   */
  fallback?: ReactNode;
  
  /**
   * Optional callback to be notified when an error occurs
   */
  onError?: (error: Error, errorInfo: ErrorInfo) => void;
  
  /**
   * Optional className to apply to the error boundary container
   */
  className?: string;
  
  /**
   * Reset error state when this key changes
   * Useful for resetting the error boundary when a significant prop changes
   */
  resetKey?: string | number;
}

interface ErrorBoundaryState {
  hasError: boolean;
  error: Error | null;
}

/**
 * ErrorBoundary component that catches JavaScript errors in its child component tree.
 * It logs the errors and displays a fallback UI instead of crashing the entire app.
 * 
 * @example Basic usage
 * ```tsx
 * <ErrorBoundary>
 *   <MyComponent />
 * </ErrorBoundary>
 * ```
 * 
 * @example With custom fallback UI
 * ```tsx
 * <ErrorBoundary
 *   fallback={<div>Something went wrong. <button onClick={refresh}>Try again</button></div>}
 *   onError={(error) => logErrorToService(error)}
 * >
 *   <MyComponent />
 * </ErrorBoundary>
 * ```
 */
class ErrorBoundary extends Component<ErrorBoundaryProps, ErrorBoundaryState> {
  constructor(props: ErrorBoundaryProps) {
    super(props);
    this.state = { hasError: false, error: null };
  }

  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    // Update state so the next render will show the fallback UI
    return { hasError: true, error };
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo): void {
    // Log the error to the console
    console.error('Error caught by boundary:', error, errorInfo);
    
    // Call the optional onError callback if provided
    if (this.props.onError) {
      this.props.onError(error, errorInfo);
    }
  }

  componentDidUpdate(prevProps: ErrorBoundaryProps): void {
    // Reset the error state when the resetKey changes
    if (this.state.hasError && this.props.resetKey !== prevProps.resetKey) {
      this.setState({ hasError: false, error: null });
    }
  }

  render(): ReactNode {
    if (this.state.hasError) {
      // Render the fallback UI if provided, otherwise use a default error message
      if (this.props.fallback) {
        return this.props.fallback;
      }
      
      return (
        <div className={`error-boundary ${this.props.className || ''}`}>
          <h2>Something went wrong</h2>
          <details>
            <summary>Error details</summary>
            <pre>{this.state.error?.toString()}</pre>
          </details>
          <button
            onClick={() => this.setState({ hasError: false, error: null })}
            style={{ marginTop: '1rem' }}
          >
            Try again
          </button>
        </div>
      );
    }

    return this.props.children;
  }
}

export default ErrorBoundary;

