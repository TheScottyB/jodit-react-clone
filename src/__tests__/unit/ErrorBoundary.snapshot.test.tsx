import React from 'react';
import { render } from '@testing-library/react';
import ErrorBoundary from '../../components/ErrorBoundary';

/**
 * Snapshot tests for ErrorBoundary component
 * 
 * Strategy:
 * 1. Test default error rendering
 * 2. Test custom fallback UI
 * 3. Test with different prop combinations
 * 4. Test error states with various error types
 */

const ErrorComponent = ({ message }: { message: string }) => {
  throw new Error(message);
  return null;
};

describe('ErrorBoundary Snapshots', () => {
  // Suppress console.error for cleaner test output
  beforeAll(() => {
    jest.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should match snapshot when no error occurs', () => {
    const { container } = render(
      <ErrorBoundary>
        <div>Normal content</div>
      </ErrorBoundary>
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with default error UI', () => {
    const { container } = render(
      <ErrorBoundary>
        <ErrorComponent message="Test error" />
      </ErrorBoundary>
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with custom fallback UI', () => {
    const customFallback = (
      <div className="custom-error">
        <h3>Custom Error View</h3>
        <p>Something went wrong in the application</p>
      </div>
    );

    const { container } = render(
      <ErrorBoundary fallback={customFallback}>
        <ErrorComponent message="Test error" />
      </ErrorBoundary>
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with custom className', () => {
    const { container } = render(
      <ErrorBoundary className="custom-boundary-class">
        <ErrorComponent message="Test error" />
      </ErrorBoundary>
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with complex error message', () => {
    const complexError = new Error('Complex error');
    complexError.stack = 'Error: Complex error\n    at Component\n    at ErrorBoundary';
    
    const ThrowComplexError = () => {
      throw complexError;
      return null;
    };

    const { container } = render(
      <ErrorBoundary>
        <ThrowComplexError />
      </ErrorBoundary>
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with nested error boundaries', () => {
    const { container } = render(
      <ErrorBoundary className="outer">
        <div>
          <ErrorBoundary className="inner">
            <ErrorComponent message="Nested error" />
          </ErrorBoundary>
        </div>
      </ErrorBoundary>
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot when error boundary itself errors', () => {
    const BrokenFallback = () => {
      throw new Error('Fallback error');
      return null;
    };

    const { container } = render(
      <ErrorBoundary fallback={<BrokenFallback />}>
        <ErrorComponent message="Component error" />
      </ErrorBoundary>
    );
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with all props provided', () => {
    const onError = jest.fn();
    const customFallback = <div>Custom error view</div>;

    const { container } = render(
      <ErrorBoundary
        fallback={customFallback}
        onError={onError}
        className="test-class"
        resetKey="123"
      >
        <ErrorComponent message="Test error" />
      </ErrorBoundary>
    );
    expect(container).toMatchSnapshot();
  });
});

