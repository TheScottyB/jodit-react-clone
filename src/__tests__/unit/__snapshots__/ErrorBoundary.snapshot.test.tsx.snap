// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Snapshots should match snapshot when no error occurs 1`] = `
<div>
  <div>
    Normal content
  </div>
</div>
`;

exports[`ErrorBoundary Snapshots should match snapshot with all props provided 1`] = `
<div>
  <div>
    Custom error view
  </div>
</div>
`;

exports[`ErrorBoundary Snapshots should match snapshot with complex error message 1`] = `
<div>
  <div
    class="error-boundary "
  >
    <h2>
      Something went wrong
    </h2>
    <details>
      <summary>
        Error details
      </summary>
      <pre>
        Error: Complex error
      </pre>
    </details>
    <button
      style="margin-top: 1rem;"
    >
      Try again
    </button>
  </div>
</div>
`;

exports[`ErrorBoundary Snapshots should match snapshot with custom className 1`] = `
<div>
  <div
    class="error-boundary custom-boundary-class"
  >
    <h2>
      Something went wrong
    </h2>
    <details>
      <summary>
        Error details
      </summary>
      <pre>
        Error: Test error
      </pre>
    </details>
    <button
      style="margin-top: 1rem;"
    >
      Try again
    </button>
  </div>
</div>
`;

exports[`ErrorBoundary Snapshots should match snapshot with custom fallback UI 1`] = `
<div>
  <div
    class="custom-error"
  >
    <h3>
      Custom Error View
    </h3>
    <p>
      Something went wrong in the application
    </p>
  </div>
</div>
`;

exports[`ErrorBoundary Snapshots should match snapshot with default error UI 1`] = `
<div>
  <div
    class="error-boundary "
  >
    <h2>
      Something went wrong
    </h2>
    <details>
      <summary>
        Error details
      </summary>
      <pre>
        Error: Test error
      </pre>
    </details>
    <button
      style="margin-top: 1rem;"
    >
      Try again
    </button>
  </div>
</div>
`;

exports[`ErrorBoundary Snapshots should match snapshot with nested error boundaries 1`] = `
<div>
  <div>
    <div
      class="error-boundary inner"
    >
      <h2>
        Something went wrong
      </h2>
      <details>
        <summary>
          Error details
        </summary>
        <pre>
          Error: Nested error
        </pre>
      </details>
      <button
        style="margin-top: 1rem;"
      >
        Try again
      </button>
    </div>
  </div>
</div>
`;
